### ISSUE TITLE
can not deserialize TypeInformation, which serialized in XCDR

### ISSUE THREAD

**Issue by binbowang1987** [eclipse-cyclonedds/cyclonedds#1572]
In SEDP message, reader anounced that it supports XCDR and XCDR2.
![图片](https://user-images.githubusercontent.com/8586061/219262762-591cb8f0-331a-485d-b492-8fcf2cad7fb1.png)

But, when look into the `TypeInformation` deserialization code. CycloneDDS only support desialize `TypeInformation` in CDR2.
https://github.com/eclipse-cyclonedds/cyclonedds/blob/011f80a68e3b44fdcc07b0065473f4cbfa5470a5/src/core/ddsi/src/ddsi_plist.c#L781-L808

**Comment by eboasson** [comment]
That's because they concern different things. `DATA_REPRESENTATION` deals with the encodings the reader accepts for *application* data, but `TypeInformation` is not application data but internal meta-data.

The `TypeInformation` is simply spec'd as little-endian XCDR2:

> 7.3.4.5 TypeObject serialization
>
> The serialization of a TypeObject shall happen in accordance to its IDL declaration and the general serialization rules defined in this specification (see Clause 7.4) for XCDR encoding version 2. Additional restrictions are placed such that the serialized result is bitwise identical independently of the vendor or platform where the serialization occurs. Specifically:
> * The serialization shall use Little Endian encoding.
> * ...

Interesting screenshot by the way: I hadn't expected to see screenshots of Wireshark that don't recognise parameter id 0x75 as `TypeInformation`, nor can I recall having seen a "Compression Id Mask" before.

And last but not least, would you be willing to tell me which DDS implementation this is coming from? I like to keep track of implementations that I know Cyclone interoperates with, and I can't see how this could've come out of Cyclone: lifespan wouldn't be part of a reader endpoint description. 😁

(And besides, ∞ would be the default anyway, and so Cyclone won't publish it unless it is configured to publish all QoS settings, but then it doesn't fit the order I would expect.)

**Comment by binbowang1987** [comment]
The DDS Implementation in wireshark is CycloneDDS(0.10.2).
Recently, I am comparing the implementation of XTypes(1.2) between CycloneDDS and FastDDS. Find some diffrences, FastDDS use `PL_CDR_LE` to encode TypeLookupService related requests, and CycloneDDS uses `CDR2_LE`. FastDDS also use CDR1 to encode TypeInformation. And CycloneDDS can not interoerate with FastDDS through XTypes(1.2)
I am doubting which one is the correct implementation.

**Comment by eboasson** [comment]
> The DDS Implementation in wireshark is CycloneDDS(0.10.2).

😱 I better go find out how that `LIFESPAN` thing showed up there, then. Thanks for helping uncover it ☺️

> Recently, I am comparing the implementation of XTypes(1.2) between CycloneDDS and FastDDS. Find some diffrences, FastDDS use PL_CDR_LE to encode TypeLookupService related requests, and CycloneDDS uses CDR2_LE. FastDDS also use CDR1 to encode TypeInformation. And CycloneDDS can not interoerate with FastDDS through XTypes(1.2)

That sounds very much like FastDDS simply doesn't implement current XTypes (and possibly not XTypes at all). We don't support the deprecated type lookup mechanisms and representations of the early, broken XTypes specs. That makes no sesne ...

We do interoperate with OpenDDS and RTI Connext — which is not to say that interoperability issues there do not exist, but they are more subtle than what you describe. The issues that do get uncovered get fixed, either by us or by the others.

> I am doubting which one is the correct implementation.

Clearly, I am pretty sure that Cyclone is "correct" and Fast-DDS is the one that's wrong here 🙂

**Comment by binbowang1987** [comment]
Yeah, I got it. Thank you for your explanation. I will continue to dig into XTypes(v1.3).

Here is the pcap file, hope to give some help.
[dds-xtypes-dynsub.zip](https://github.com/eclipse-cyclonedds/cyclonedds/files/10764482/dds-xtypes-dynsub.zip)
